Add support for CoqMakefile.local-late (#12411)
coq/coq: doc/sphinx/practical-tools/utilities.rst

**Variable assignment**

> The variable may belong to the variables listed in the ``Parameters``

Author: @gares Given the comments you made, is this change incorrect?

Reviewer: yes

discouraged, since the contents of this file, including variable names and
status of rules, may change in the future.

Use the optional file ``CoqMakefile.local`` to extend

> ``CoqMakefile``. In particular, you can declare custom actions to be
> performed before or after the build process. Similarly you can

Reviewer 2: ```suggestion
``CoqMakefile``. In particular, you can declare custom actions to
perform before or after the build process. Similarly you can
```

Author: Nope, this one changes the meaning in a way that sounds wrong.  The make program performs the actions declared.  The actions don't perform anything.  Replacing "be performed" with "run" would work, if you think that's better.

Reviewer 2: "run" may be better

Author: ```suggestion
``CoqMakefile``. In particular, you can declare custom actions to
run before or after the build process. Similarly you can
```

customize the install target or even provide new targets. See
:ref:`coqmakefilelocal` for extension-point documentation. To access
variables defined in the ``Makefile`` in a way not covered by
`secondary expansion
<https://www.gnu.org/software/make/manual/html_node/Secondary-Expansion.html>`_,
for example to add new dependencies to compiled outputs, use the

> optional file ``CoqMakefile.local-late``.  See
> :ref:`coqmakefilelocallate` for a non-exhaustive list.

Reviewer 2: ```suggestion
optional file ``CoqMakefile.local-late``.  See
:ref:`coqmakefilelocallate` for a non-exhaustive list of variables.
```

> The extensions of the files listed in ``_CoqProject`` are used in order to
> decide how to build them. In particular:

Reviewer 2: ```suggestion
The extensions of files listed in ``_CoqProject`` determine
how they are built. In particular:
```

extend ``CoqMakefile``. In particular one can declare custom actions to be
performed before or after the build process. Similarly one can customize the
install target or even provide new targets. Extension points are documented in

> paragraph :ref:`coqmakefilelocal`. If the user needs to access variables
> defined in the ``Makefile`` in a way not covered by `secondary expansion
> <https://www.gnu.org/software/make/manual/html_node/Secondary-Expansion.html>`_,
> for example to add new dependencies to compiled outputs, the optional file
> ``CoqMakefile.local-late`` can be used.  A non-exhaustive list is documented in
> paragraph :ref:`coqmakefilelocallate`.

Reviewer 2: ```suggestion
install target or even provide new targets. See
:ref:`coqmakefilelocal`. To access variables
defined in the ``Makefile`` in a way not covered by `secondary expansion
<https://www.gnu.org/software/make/manual/html_node/Secondary-Expansion.html>`_,
for example to add new dependencies to compiled outputs, use the optional file
``CoqMakefile.local-late``.  See :ref:`coqmakefilelocallate` for a non-exhaustive list.
```

Author: This change results in a switch from passive voice to active voice in the middle of the paragraph.  I'm inclined to leave the addition as-is here.

Author: (Also, the change you suggest makes it less clear, I think, what explanation the reader should expect by following the paragraph references.)

Reviewer 2: I will leave it up to you.

Nonetheless, while the passive voice is commonly used in academic writing, we're aiming for a more readable style.  The preceding sentences would read better as:

> Use the optional file ``CoqMakefile.local`` to
extend ``CoqMakefile``. In particular, you can declare custom actions to
perform before or after the build process.

This is wordy: "If the user needs to access variables"
Hyperlinks should just be "See ..."; "documented in paragraph :ref:`coqmakefilelocallate`." is wordy: everything in the file is documentation and it's not necessary to say that it's a paragraph.I will leave it up to you.

Nonetheless, while the passive voice is commonly used in academic writing, we're aiming for a more readable style.  The preceding sentences would read better as:

> Use the optional file ``CoqMakefile.local`` to
extend ``CoqMakefile``. In particular, you can declare custom actions to
perform before or after the build process.

This is wordy: "If the user needs to access variables"
Hyperlinks should just be "See ..."; "documented in paragraph :ref:`coqmakefilelocallate`." is wordy: everything in the file is documentation and it's not necessary to say that it's a paragraph.

Author: Updated, please take a look

extend ``CoqMakefile``. In particular one can declare custom actions to be
performed before or after the build process. Similarly one can customize the
install target or even provide new targets. Extension points are documented in
paragraph :ref:`coqmakefilelocal`. If the user needs to access variables

> defined in the ``Makefile``, the optional file ``CoqMakefile.local-late`` can

Reviewer: I'd like to a have a *non exhaustive* list of variables which are accessible, a bit like the ones below. Ideally one should not read the generated file in order to write the .local one.

Author: I've added some more documentation, please take a look and let me know what you think.

**Variable assignment**

The variable must belong to the variables listed in the ``Parameters``

> section of the generated makefile.  If access to more variables are
> needed, ``CoqMakefile.local-late`` can be used instead. Here we
> describe only few of the variables.

Reviewer 2: ```suggestion
section of the generated makefile.  Use 
``CoqMakefile.local-late`` instead to access more variables. These include:
```

  One can append lines to the generated ``.merlin`` file extending this
  target.

.. _coqmakefilelocallate:

CoqMakefile.local-late
++++++++++++++++++++++

The optional file ``CoqMakefile.local-late`` is included at the end of the generated

> file ``CoqMakefile``.  The following is an incomplete list of accessible variables:

Reviewer 2: ```suggestion
file ``CoqMakefile``.  The following is an partial list of accessible variables:
```

Author: The "an" needs to change too then
```suggestion
file ``CoqMakefile``.  The following is a partial list of accessible variables:
```

> :COQ_VERSION:

Reviewer: This variable is also available in the non `-late` file (I'm using it already). Please move it upward.
Ditto for the the next one.

Author: All variables are available for use in *recipes* of rules in the non-late file, as well as in the definitions of variables assigned using `=`.  They are also available in the prerequisites of rules only if you double the `$` and enable secondary expansion.  However, these variables are not available when naming the outputs of rules, nor in top-level `ifeq`/`ifneq` statements, nor in variables assigned with `:=`.  (I presume your use-cases involve the former kind and not the latter kind?)  Do you have suggestions for making this distinction more clear in the documentation?

Reviewer: Indeed. I think it is fine, your explanation could also go in the doc, since the external pointer is a bit boring to follow.

Author: I've added some explanation in the doc, please take a look

  One can append lines to the generated ``.merlin`` file extending this
  target.

.. _coqmakefilelocallate:

CoqMakefile.local-late
++++++++++++++++++++++

The optional file ``CoqMakefile.local-late`` is included at the end of the generated
file ``CoqMakefile``.  The following is an incomplete list of accessible variables:

:COQ_VERSION:

>    describes the version of ``coqc`` being used, and can be used to
>    provide different behavior depending on what version of Coq is
>    being used

Reviewer 2: ```suggestion
   the version of ``coqc`` being used, which can be used to
   provide different behavior depending on the Coq version
```

:COQMAKEFILE_VERSION:

>    describes the version of Coq that was used to generate the
>    Makefile, and can be used to detect version mismatches

Reviewer 2: ```suggestion
   the version of Coq that used to generate the
   Makefile, which can be used to detect version mismatches
```

Author: You forgot to delete `that` before `was`, but thanks!
```suggestion
   the version of Coq used to generate the
   Makefile, which can be used to detect version mismatches
```

:ALLDFILES:

>    contains the list of generated dependency files, and can be used,
>    for example, to instruct ``make`` to recompute all dependencies
>    when some file changes by writing ``$(ALLDFILES): myfile`` or to
>    indicate that some file must be generated before dependencies can
>    be computed by writing ``$(ALLDFILES): | mygeneratedfile``

Reviewer 2: ```suggestion
   the list of generated dependency files, which can be used,
   for example, to cause ``make`` to recompute dependencies
   when one or more files change by writing ``$(ALLDFILES): myfile`` or to
   indicate that a file must be generated before dependencies can
   be computed by writing ``$(ALLDFILES): | mygeneratedfile``
```

Author: No reason to explicitly say "one or more"
```suggestion
   the list of generated dependency files, which can be used,
   for example, to cause ``make`` to recompute dependencies
   when files change by writing ``$(ALLDFILES): myfiles`` or to
   indicate that files must be generated before dependencies can
   be computed by writing ``$(ALLDFILES): | mygeneratedfiles``
```

:VOFILES, GLOBFILES, CMOFILES, CMXFILES, OFILES, CMAFILES, CMXAFILES, CMIFILES, CMXSFILES:
   lists of files that are generated by various invocations of the compilers

Additionally, the following varaibles may be useful for running code
to decide what targets to present via ``$(shell ...)``; these

> variables are already accessible in the recipe for any rules added in

Author: ```suggestion
variables are already accessible in the recipes for any rules added in
```

> Additionally, the following varaibles may be useful for running code
> to decide what targets to present via ``$(shell ...)``; these
> variables are already accessible in the recipe for any rules added in
> ``CoqMakefile.local``, but are only accessible to top-level ``$(shell
> ...)`` invocations in ``CoqMakefile.local-late``:

Reviewer 2: ```suggestion
In addition, the following variables may be useful for
to decide what targets to present via ``$(shell ...)``; these
variables are already accessible in the recipe for any rules added in
``CoqMakefile.local``, but are only accessible to top-level ``$(shell
...)`` invocations in ``CoqMakefile.local-late``:
```

"in the recipe"?  Seems imprecise/informal, is there better wording?

Author: Perhaps it should be "in the recipes"?  "Recipe" and "rule" are technical terms here, see https://www.gnu.org/software/make/manual/html_node/Recipes.html#Recipes

Reviewer 2: Or "in recipes for rules"

Also "accessible to" -> "accessible from"

Author: Done
